src/ui/public/vislib/lib/axis.js

```diff
@@ -0,0 +1,228 @@
+import d3 from 'd3';
+import _ from 'lodash';
+import $ from 'jquery';
+import ErrorHandlerProvider from 'ui/vislib/lib/_error_handler';
+import AxisTitleProvider from 'ui/vislib/lib/axis_title';
+import AxisLabelsProvider from 'ui/vislib/lib/axis_labels';
+import AxisScaleProvider from 'ui/vislib/lib/axis_scale';
+
+export default function AxisFactory(Private) {
+  const ErrorHandler = Private(ErrorHandlerProvider);
+  const AxisTitle = Private(AxisTitleProvider);
+  const AxisLabels = Private(AxisLabelsProvider);
+  const AxisScale = Private(AxisScaleProvider);
+  const defaults = {
+    show: true,
+    type: 'value',
+    elSelector: '.axis-wrapper-{pos} .axis-div',
+    position: 'left',
+    axisFormatter: null, // TODO: create default axis formatter
+    scale: 'linear',
+    expandLastBucket: true, //TODO: rename ... bucket has nothing to do with vis
+    inverted: false,
+    style: {
+      color: '#ddd',
+      lineWidth: '1px',
+      opacity: 1,
+      tickColor: '#ddd',
+      tickWidth: '1px',
+      tickLength: '6px'
+    }
+  };
+
+  const categoryDefaults = {
+    type: 'category',
+    position: 'bottom',
+    labels: {
+      rotate: 0,
+      rotateAnchor: 'end',
+      filter: true
+    }
+  };
+  /**
+   * Appends y axis to the visualization
+   *
+   * @class Axis
+   * @constructor
+   * @param args {{el: (HTMLElement), yMax: (Number), _attr: (Object|*)}}
+   */
+  class Axis extends ErrorHandler {
+    constructor(args) {
+      super();
+      if (args.type === 'category') {
+        _.extend(this, defaults, categoryDefaults, args);
+      } else {
+        _.extend(this, defaults, args);
+      }
+
+      this._attr = args.vis._attr;
+      this.elSelector = this.elSelector.replace('{pos}', this.position);
+      this.scale = new AxisScale(this, {scale: this.scale});
+      this.axisTitle = new AxisTitle(this, this.axisTitle);
```

Comment:
this.axisTitle ?

what happens is that i would get in configuration object .... now i am replacing parts of it with real class instances ...

so before this.axisTitle is config object for axisTitle ... and then i create an AxisTitle instance with that same config.